honeyflare/__init__.py

```diff
@@ -1,2 +1,197 @@
+    :param lock_bucket: If you want to use a dedicated bucket for holding locks
+        and completion status, pass it here. Otherwise the bucket that holds the
+        logs will be used (requires write access to that bucket).
+    '''
+    if sampling_rate_by_status is None:
+        sampling_rate_by_status = {}
+
+    if lock_bucket is None:
+        lock_bucket = bucket
+
+    compiled_patterns = [compile_pattern(p) for p in patterns or []]
+    libhoney_client = create_libhoney_client(honeycomb_key, honeycomb_dataset)
+
+    lock = GCSLock(lock_bucket, 'locks/%s' % object_name)
+    total_events = 0
+    with lock:
+        if is_already_processed(lock_bucket, object_name):
+            # We might have been retried due to a failure but another
+            # function succeeded in the meantime
+            return
+
+        local_path = download_file(bucket, object_name)
+
+        for sample_rate, entry in get_sampled_file_entries(local_path, sampling_rate_by_status):
+            event = libhoney_client.new_event()
+            event.sample_rate = sample_rate
+            enrich_entry(entry, compiled_patterns, query_param_filter)
+            event.add(entry)
+            event.created_at = datetime.datetime.utcfromtimestamp(entry['EdgeEndTimestamp']/1e9)
+            event.send_presampled()
+            total_events += 1
+
+        libhoney_client.close()
+        os.remove(local_path)
+        mark_as_processed(lock_bucket, object_name)
+    return total_events
+
+
+def create_libhoney_client(writekey, dataset):
+    client = libhoney.Client(
+        writekey=writekey,
+        dataset=dataset,
+        block_on_send=True,
+        user_agent_addition='honeyflare/%s' % __version__,
+    )
+    client.add_field('MetaProcessor', 'honeyflare/%s' % __version__)
+
+    thread = threading.Thread(target=read_honeycomb_responses, args=(client.responses(), dataset))
+    thread.start()
+
+    return client
+
+
+def read_honeycomb_responses(resp_queue, dataset):
+    '''Log failing responses from honeycomb'''
+    while True:
+        resp = resp_queue.get()
+        if resp is None:
+            # The client will enqueue a None value after we call client.close()
+            break
+
+        if resp['status_code'] > 400:
+            sys.stderr.write('Got %d from honeycomb when submitting to %s: %s\n' % (
+                resp['status_code'], dataset, resp['error']))
+
+
+def is_already_processed(lock_bucket, object_name):
+    return _processed_blob(lock_bucket, object_name).exists()
+
+
+def mark_as_processed(lock_bucket, object_name):
+    blob = _processed_blob(lock_bucket, object_name)
+    try:
+        blob.upload_from_string(b'')
+    except PreconditionFailed:
+        # Another invocation has already processed this but it failed to be
+        # caught in the lock. Ignore
+        pass
+
+
+def _processed_blob(bucket, object_name):
+    return bucket.blob('completed/%s' % object_name)
```

Comment:
I'm guessing this is creating a file in the bucket in the completed directory, would it make sense to use something more ephemeral like a redis key expiring after 24 hours?

Reply:
I guess keeping everything scoped to gcloud makes it easier at not having to worry about dependencies

Reply:
Yup, that was the idea, no extra services managed by us.